### ISSUE TITLE
datacenter, cluster, resource_pool, mac_addresses details missing

### ISSUE THREAD

**Issue by plribeiro3000** [fog/fog-vsphere#47]
When I issue a list_virtual_machines, the datacenter, cluster, resource_pool, mac_addresses fields seem to not come across properly:

```
 {"id"=>"5029f2a5-6162-025b-7da4-a8790961f0ed",
  "name"=>"cloned-ubuntu",
  "uuid"=>"4229703c-388e-f28d-4a49-5b263cd3b633",
  "template"=>false,
  "parent"=>Folder("group-v69"),
  "hostname"=>nil,
  "operatingsystem"=>nil,
  "ipaddress"=>nil,
  "power_state"=>"poweredOff",
  "connection_state"=>"connected",
  "hypervisor"=>
   #<Proc:0x000000038a2390@/usr/lib64/ruby/gems/1.9.1/gems/fog-1.22.0/lib/fog/vsphere/compute.rb:150>,
  "tools_state"=>"toolsNotInstalled",
  "tools_version"=>"guestToolsNotInstalled",
  "memory_mb"=>1024,
  "cpus"=>1,
  "corespersocket"=>1,
  "overall_status"=>"green",
  "guest_id"=>"ubuntu64Guest",
  "mo_ref"=>"vm-79",
  "datacenter"=>
   #<Proc:0x000000038a2480@/usr/lib64/ruby/gems/1.9.1/gems/fog-1.22.0/lib/fog/vsphere/compute.rb:148>,
  "cluster"=>
   #<Proc:0x000000038a2408@/usr/lib64/ruby/gems/1.9.1/gems/fog-1.22.0/lib/fog/vsphere/compute.rb:149>,
  "resource_pool"=>
   #<Proc:0x000000038a2340@/usr/lib64/ruby/gems/1.9.1/gems/fog-1.22.0/lib/fog/vsphere/compute.rb:151>,
  "mac_addresses"=>
   #<Proc:0x000000038a22c8@/usr/lib64/ruby/gems/1.9.1/gems/fog-1.22.0/lib/fog/vsphere/compute.rb:155>,
  "path"=>"/Datacenters/testdc/vm/New VMs",
  "relative_path"=>"testdc/New VMs"}]
```

Original issue opened by @stonith at fog/fog#2950

**Comment by plribeiro3000** [comment]
cc/ @geemus @nirvdrum

**Comment by jrgarcia** [comment]
The problem is clearly right [here](https://github.com/fog/fog-vsphere/blob/master/lib/fog/vsphere/compute.rb#L182-L185). I'll take a look at fixing this today or tomorrow.

**Comment by jrgarcia** [comment]
Looking at a blame on this code points to fog/fog@036a0370744002d6727e68fdf075a3a3adca2559. Unfortunately, without any context, I'm unsure of why this change was originally made.

**Comment by plribeiro3000** [comment]
Well. The bad part is, there is no PR for this change so the real reason behind that might not come to light again unless @aussielunix or @jeffmccune chime in since this was on a really old version of `fog` - `1.1.11`.

It seems the issue was `allowing ram and num of cpu to be set when cloning` that started at fog/fog#1732.

I'm completely out on context here so i will just throw some ideas out. 

What if we don't return those attributes out of the box? IF you need/want you have to make an extra call for each object and/or we could have another method that does a blocking call.
What do you think?

**Comment by jrgarcia** [comment]
@plribeiro3000 [This code](https://github.com/fog/fog/blob/036a0370744002d6727e68fdf075a3a3adca2559/lib/fog/vsphere/models/compute/server.rb#L55-L62) was added with the `Proc` stuff and appears to handle that. The issue I see, is if someone wants to use that `Hash` for reporting or something, they don't have that info available in the `Hash` without a `call` to those `Proc`s first.

**Comment by plribeiro3000** [comment]
I see. 

I guess my point is that on the scenario you mentioned the user will have to wait a blocking operation anyway. So maybe we might want to remove that lazy loading feature on that call and remove support for those attributes (on this call). And then create a new one that would be a blocking operation and return everything.

Do you think thats something we can do without breaking too much clients of `fog-vsphere`?

**Comment by jrgarcia** [comment]
I'm not sure. Let me think on this a bit. That was my thought too, but I really don't want to break existing usage.

**Comment by plribeiro3000** [comment]
I agree. My thought right now is that maybe this breaking now will be better then a bigger breaking that might occur later if we don't act now. But i can't measure this, only you (😬).
Now that you pointed the attribute getters being overwritten to work lie that i believe this won't be a big issue later. The only thing to note is that it is overwriting the default behaviour of an attribute getter which will will drop any new support or bugfix we do on `fog-core`.

**Comment by plribeiro3000** [comment]
What about pushing lazy loading evaluated attributes to `fog-core` instead and then use it here as a DSL?

**Comment by jrgarcia** [comment]
Do you mean in general or does `fog-core` already support lazy loading?

**Comment by plribeiro3000** [comment]
I don't think it does, but we can add it. I'm pretty sure other providers could improve with such feature.

**Comment by jrgarcia** [comment]
Ok. If that would be useful for others, I think that would be a good option. If it doesn't benefit the majority of providers, I'd say it doesn't belong in there (which I'm sure all would agree with). I'm up for discussing that further.

**Comment by geemus** [comment]
I think there may be some lazy-loading in other places, but done manually (I don't think we've tried to do a more universal abstraction yet).
